File: 0020-valid-parentheses/0020-valid-parentheses.cpp
#include <stack>
class Solution {
public:
    bool isValid(string s) {
        std::vector<char> open;
        for (auto letter : s) {
            if( letter=='{' || letter=='[' || letter=='(') {
                open.push_back(letter);   
            } else {
                if(open.empty())
                    return false;
                
                switch(letter){
                    case '}':
                        if(open.back() != '{')
                           return false;
                        break;
                    case ']':
                        if(open.back() != '[')
                           return false;
                         break;
                    case ')': 
                        if(open.back() != '(')
                           return false;
                        break;
                }
                open.pop_back();
            }
        }
        return open.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        options = {"(":")","{":"}","[":"]"}
        close = {"}","]",")"}
        for letter in s:
            if letter in options:
                stack.append(letter)
            if letter in close:
                if len(stack) == 0:
                    return False
                a = stack.pop()
                if options[a] != letter:
                    return False
        if len(stack) != 0:
            return False    
        return True

